pkg/keyidentifier/keyidentifier.go

```diff
@@ -0,0 +1,185 @@
+}
+
+func (kIdentifer *KeyIdentifier) IdentifyFile(path string) (*KeyInfo, error) {
+	level.Debug(kIdentifer.logger).Log(
+		"msg", "starting a key identification",
+		"file", path,
+	)
+
+	keyBytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "read key")
+	}
+
+	ki, err := kIdentifer.Identify(keyBytes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "reading file %s", path)
+	}
+
+	return ki, nil
+}
+
+// Identify uses a manually curated set of heuristics to determine
+// what kind of key something is. Generally speaking, we consider `err
+// == nil` as success, and throw away errors as unparsable keys.
+func (kIdentifer *KeyIdentifier) Identify(keyBytes []byte) (*KeyInfo, error) {
+	level.Debug(kIdentifer.logger).Log(
+		"msg", "starting a key identification",
+		"file", "<bytestream>",
+	)
+
+	// Some magic strings for dispatching. These are the simplest.
+	switch {
+	case bytes.HasPrefix(keyBytes, []byte(ppkBegin)):
+		return ParsePuttyPrivateKey(keyBytes)
+	case bytes.HasPrefix(keyBytes, []byte(sshcomBegin)):
+		return ParseSshComPrivateKey(keyBytes)
+	case bytes.HasPrefix(keyBytes, []byte(ssh1LegacyBegin)):
+		return ParseSsh1PrivateKey(keyBytes)
+	}
+
+	// If nothing else fits. treat it like a pem
+	if ki, err := kIdentifer.attemptPem(keyBytes); err == nil {
+		return ki, nil
+	}
+
+	// Out of options
+	return nil, errors.New("Unable to parse key")
+
+}
+
+// attemptPem tries to decode the pem, and then work with the key. It's
+// based on code from x/crypto's ssh.ParseRawPrivateKey, but more
+// flexible in handling encryption and formats.
+func (kIdentifer *KeyIdentifier) attemptPem(keyBytes []byte) (*KeyInfo, error) {
+	ki := &KeyInfo{
+		Format: "",
+		Parser: "attemptPem",
+	}
+
+	block, _ := pem.Decode(keyBytes)
+	if block == nil {
+		return nil, errors.New("pem could not parse")
+	}
+
+	ki.Encrypted = boolPtr(pemEncryptedBlock(block))
+
+	level.Debug(kIdentifer.logger).Log(
+		"msg", "pem decoded",
+		"block type", block.Type,
+	)
+
+	switch block.Type {
+	case "RSA PRIVATE KEY":
+		ki.Type = ssh.KeyAlgoRSA
+		ki.Format = "openssh"
+
+		if key, err := x509.ParsePKCS1PrivateKey(block.Bytes); err == nil {
+			ki.Bits = len(key.PublicKey.N.Bytes()) * 8
+		}
+
+		return ki, nil
+
+	case "PRIVATE KEY":
+		// RFC5208 - https://tools.ietf.org/html/rfc5208
+		ki.Encrypted = boolPtr(x509.IsEncryptedPEMBlock(block))
+		if key, err := x509.ParsePKCS8PrivateKey(block.Bytes); err == nil {
+			switch assertedKey := key.(type) {
+			case *rsa.PrivateKey:
+				ki.Bits = assertedKey.PublicKey.Size() * 8
+				ki.Type = "rsa"
+			case *ecdsa.PrivateKey:
+				ki.Bits = assertedKey.PublicKey.Curve.Params().BitSize
+				ki.Type = "ecdsa"
+			}
+		}
+		return ki, nil
+
+	case "EC PRIVATE KEY":
+		// set the Type here, since parsing fails on encrypted keys
+		ki.Type = "ecdsa"
+
+		if key, err := x509.ParseECPrivateKey(block.Bytes); err == nil {
+			ki.Type = "ecdsa"
+			ki.Bits = key.PublicKey.Curve.Params().BitSize
```

Comment:
Encrypted is set up on line 116, in `ki.Encrypted = boolPtr(pemEncryptedBlock(block))`. Setting it here seems uneeded. It would matter if this was refactored, but that's what the tests are for